counsel.el

```diff
@@ -5394,6 +5432,46 @@ specified by the `blddir' property."
             :action #'counsel-compile--action
             :caller 'counsel-compile))
 
+
+;(defun ivy--format-function-generic (selected-fn other-fn cands separator)
+(defun counsel-compile-env--format-hint (cands)
+  "Return a formatter for compile-env CANDS."
+  (ivy--format-function-generic
+   (lambda (selected)
+     (format "%s %s"
+             (if (member selected counsel-compile-env)
+                 (propertize "remove" 'face 'font-lock-warning-face)
+               (propertize "add" 'face
+                           'font-lock-variable-name-face))
+             selected))
+   (lambda (other)
+     other)
+   cands
+   "\n"))
+
+(defun counsel-compile-env--update (var)
+  "Update `counsel-compile-env' either adding or removing VAR."
+  (if (member var counsel-compile-env)
+      (setq counsel-compile-env (delete var counsel-compile-env))
+    (if (string-match-p counsel-compile-env-pattern var)
+        (add-to-list 'counsel-compile-env var)
+      (message "Did not add badly formed variable '%s' to environment" var))))
```

Comment:
It's more Emacs-y to use `cond` in place of nested `if`s, and it is better to use `push` or `cl-pushnew` instead of `add-to-list` in Elisp programs. I also wonder if the `message` will be annoying or unnoticed? Would `lwarn` or `user-error` be better?

```el
(cond ((member var counsel-compile-env)
       (setq counsel-compile-env (delete var counsel-compile-env)))
      ((string-match-p counsel-compile-env-pattern var)
       (push var counsel-compile-env))
      (t (message "Ignoring malformed variable: '%s'" var)))
```

Reply:
I did have warn but that was far too in your face. I'll check out lwarn and user error.

Reply:
I did have warn but that was far too in your face. I'll check out lwarn and user error.

Reply:
`lwarn` and `warn` are the same, except for the fact that packages should never use `warn`, as it doesn't let you specify a custom group TYPE.

Reply:
OK - I went for user-error, warn/lwarn is just too distracting because it pops up a new window.

Reply:
Of course, yet another option is to add validation to the `defcustom` (see [`(elisp) Type Keywords`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Type-Keywords.html)) and/or just ignore malformed entries without warning the user, the choice is yours.